### ISSUE TITLE
SELECT/SET multiple cards causes offset pile

### ISSUE THREAD

**Issue by bjalder26** [ArnoldSmith86/virtualtabletop#783]
SELECTing then SETting more than one cards produces a weirdly placed pile holder because although the cards will form a pile, and will be in the spot signified in the JSON, the pile will be at 0,0 and the cards will have the specified x/y values, so they are offset from the pile, which is at 0,0. 

Discovered when testing #782

**Comment by rogerl50** [comment]
Can you post a vtt file exhibiting the problem? Unfortunately #782 was merged and the room is no longer available.

**Comment by ArnoldSmith86** [comment]
PR test server rooms get copied to production when merging: https://virtualtabletop.io/PR782-pr-test

Or am I missing something?

**Comment by rogerl50** [comment]
> PR test server rooms get copied to production when merging: https://virtualtabletop.io/PR782-pr-test
> 
> 
> 
> Or am I missing something?

I had no idea that was the case!

**Comment by rogerl50** [comment]
A simpler version of this behavior is exhibited by putting a few cards together, forming a pile `pileid`, then clicking a button with the following `clickRoutine`:
```
  "clickRoutine": [
    {
      "func": "SELECT",
      "property": "parent",
      "value": "pileid"
    },
    {
      "func": "FOREACH",
      "collection": "DEFAULT",
      "loopRoutine": [
        {
          "func": "SET",
          "collection": "DEFAULT",
          "property": "x",
          "value": 200
        }
      ]
    }
  ]
```
I'm not sure what the software is supposed to do other than what it did: it set the `x` coordinates of the pair of cards to 200 (relative to the pile location), but left their parent as is. Thus they are still "in" the pile. Should changing the `x` or `y` coordinate also remove them from the pile, as a `MOVEXY` does (or at least as it would do if it could operate on a collection rather than just a holder)?

**Comment by bjalder26** [comment]
> I'm not sure what the software is supposed to do other than what it did: it set the `x` coordinates of the pair of cards to 200 (relative to the pile location), but left their parent as is. Thus they are still "in" the pile. Should changing the `x` or `y` coordinate also remove them from the pile, as a `MOVEXY` does (or at least as it would do if it could operate on a collection rather than just a holder)?

I would have expected the cards to form a pile at the given coordinates just like if you manually moved the cards to that spot one at a time.

**Comment by rogerl50** [comment]
> I would have expected the cards to form a pile at the given coordinates just like if you manually moved the cards to that spot one at a time.

If you replace the `SET` by `MOVEXY`, that is exactly what happens. `MOVEXY` explicitly changes the parent, while `SET` does not.

**Comment by bjalder26** [comment]
I guess I'm not sure if I think SET should offset from a pile or offset from the pile's parent.  I tend to think people shouldn't have to think about piles while making games. SET also works a bit different from MOVEXY in that MOVEXY moves cards from bottom to top, whereas SET moves them from top to bottom.  So if MOVEXY moves 5 cards it will retain the same order of the cards, but by moving the bottom first and the top one last.

**Comment by rogerl50** [comment]
Well, in the case above, the pile's parent is null, so even if you think about it as offsetting from the pile parent, it does the same thing. I agree that piles are the real problem; it is reasonable to think that two cards in the same location should form a pile independent of any other considerations.

That said, I'm not sure how or if to address this particular issue.

**Comment by bjalder26** [comment]
Yeah, the cards will wind up in the same place on the board, and it's just the pile that will wind up in a different place meaning the holder appears somewhere else.  

I think if you SET 1 card at a time (null parent & x/y), they don't even form piles.  So I THINK if the pile was broken when the x/y was SET then it wouldn't even form a pile right now.  MAYBE the desired behavior for both would be to form piles at the specified x/y, but I can't say for sure.  I'm not sure if my knowledge of various games is broad enough to say.  

However, there is a part of me that likes to make things "come together" and would like to see dragging a card to a position work the same as SETting a card to a position.  That doesn't mean it is the correct choice though.

**Comment by ArnoldSmith86** [comment]
I wouldn't expect a `SET` to change any other properties. That's what `MOVEXY` is for IMHO.

Another thing to keep in mind: we probably soon have piles with offset children to keep up with features in PCIO.

Just my 2c.

**Comment by rogerl50** [comment]
I vote to close this issue.

**Comment by bjalder26** [comment]
> I wouldn't expect a `SET` to change any other properties. That's what `MOVEXY` is for IMHO.
> 
> Another thing to keep in mind: we probably soon have piles with offset children to keep up with features in PCIO.
> 
> Just my 2c.

If a routine sets the parent to null, and the x to 100 for a single card, then 1 card winds up at x=100 on the board. If a routine sets the parent to null, and the x to 100 for two cards, then 2 cards wind up at x = 100 on the board, however instead of having a null parent they wind up part of a pile at x=100.  The pile however, is at x=0, and they're offset from the pile instead of the board (same spot though, because the pile is at x=0).

I would think that either they should wind up with parent=null and no pile, or the pile as the parent, but the pile is at x=100 (just like if they were both manually dropped there).

MOVEXY will move a pile of cards to the same location, but I believe they will be in a different order.  MOXEXY will maintain the order of the cards.  In other words if it is moving 5 cards, then it will move the one that is 5th from the top first, and the one on the top last giving a stack that is in the same order.  SET will move the top card first and the bottom card last inverting the order (the same way as if you manually picked up cards one at a time).

The current way to get the same order as SETting 5 cards to the board without getting a pile at 0/0 and an offset from the pile would be to use a FOREACH loopRoutine.  In that case it will not form a pile.

**Comment by robartsd** [comment]
> MOVEXY will move a pile of cards to the same location, but they will be in a different order. MOXEXY will maintain the order of the cards. In other words if it is moving 5 cards, then it will move the one that is 5th from the top first, and the one on the top last giving a stack that is in the same order. SET will move the top card first and the bottom card last inverting the order (the same way as if you manually picked up cards one at a time).

We have had proposals to add a mode to MOVEXY (and MOVE) that would control how the cards are moved (all at once like it currently is or a specific number at a time).

**Comment by ArnoldSmith86** [comment]
> I would think that either they should wind up with parent=null and no pile, or the pile as the parent, but the pile is at x=100 (just like if they were both manually dropped there).

Yeah, that's what I would expect. I guess I didn't read the issue text well enough...

**Comment by rogerl50** [comment]
> The current way to get the same order as SETting 5 cards to the board without getting a pile at 0/0 and an offset from the pile would be to use a FOREACH loopRoutine. In that case it will not form a pile.

Are you proposing that not forming a pile in this case (as I did in the posted example above) is the correct behavior? If so, I don't understand why the proposal in this issue is to form a new pile. Seems like exactly the same behavior.

**Comment by bjalder26** [comment]
> Are you proposing that not forming a pile in this case (as I did in the posted example above) is the correct behavior? If so, I don't understand why the proposal in this issue is to form a new pile. Seems like exactly the same behavior.

Honestly, I'm not 100% sure if it "should" form a pile or not.  I suppose there are 3 things that could happen.  

1. SELECT/parent change/SETx/y sends cards to the x/y coordinate without forming piles
2. SELECT/parent change/SETx/y sends cards to the x/y coordinate and forms a pile at the x/y coordinate afterwards
3. SELECT/parent change/SETx/y creates a pile at x=0, y=0, and offsets the cards from that point (because the pile is now the parent)

I don't really see utility in the 3rd one because it will always place the pile at 0/0, which always places the handle in the same place.

If we changed it and didn't form piles, then I think the workaround to form piles would be to use a holder or a FOREACH that explicitly selects piles.

If we changed it and did have it form piles, then I think the workaround to not form piles would be to use a holder (that doesn't form piles) or to use a FOREACH that moves the cards one at a time.

I'm not sure what people would most often expect to happen, and which one would have the most utility.  I guess it also depends on what other changes might get made.

It might give the most flexibility to make SELECT/parent change/SETx/y NOT form a pile, and eventually make it so MOVEXY (which does form a pile) can be reversed.

**Comment by rogerl50** [comment]
I think the SELECT is a red herring, as is the distinction between the FOREACH and no FOREACH. If you look at https://virtualtabletop.io/PR782-pr-test, I further modified it to add a button that does what the first one did, but without the SELECT/parent change. So it just changes the x/y coordinates of the top two cards. They also do not form a pile. 

The behavior that is being exhibited here is (as @ArnoldSmith86 said a few posts back) that SET just changes the property being set, and changes nothing else. If you accept that as desirable behavior, then the real problem is (gasp) piles. We want them to be "invisible", yet as currently implemented, they are full-fledged widgets that can have children, and if SET doesn't change the parent by default in this case, then it shouldn't create a *new* pile.

So I guess I'm not advocating for either solution, but until we are clear on what SET should do, and have "solved" the pile problem, I don't think we will have an answer for this issue that is intuitive in all cases.

**Comment by bjalder26** [comment]
> I think the SELECT is a red herring, as is the distinction between the FOREACH and no FOREACH. If you look at https://virtualtabletop.io/PR782-pr-test, I further modified it to add a button that does what the first one did, but without the SELECT/parent change. So it just changes the x/y coordinates of the top two cards. They also do not form a pile.
> 
> The behavior that is being exhibited here is (as @ArnoldSmith86 said a few posts back) that SET just changes the property being set, and changes nothing else. If you accept that as desirable behavior, then the real problem is (gasp) piles. We want them to be "invisible", yet as currently implemented, they are full-fledged widgets that can have children, and if SET doesn't change the parent by default in this case, then it shouldn't create a _new_ pile.
> 
> So I guess I'm not advocating for either solution, but until we are clear on what SET should do, and have "solved" the pile problem, I don't think we will have an answer for this issue that is intuitive in all cases.

Yeah, brining up the FOREACH was meant as a way of saying there's a way to do either thing regardless.  The button you made doesn't form a pile because the cards never leave the original pile, and when you break the pile (by picking up 1 of 2 cards) the other card falls back in the holder.

I agree that Piles are the issue.  Right now SET creates piles, thus causes a parent change not explicitly called for in the JSON. What I think is probably the case is that there is no clear best way to deal with this right now.  I don't feel like I have the answer.  So this probably wait.  I think closing the issue until an apparent right answer and/or use is known would be reasonable.
